### PR TITLE
Add System.Security.Cryptography.Encryption.Aes Source And Tests

### PR DESCRIPTION
**Pull request by bartonjs**
Resolves https://github.com/dotnet/corefx/issues/1925.

### CODE REVIEW

File: src/System.Security.Cryptography.Encryption.Aes/src/System.Security.Cryptography.Encryption.Aes.csproj [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'OSX_Debug|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'OSX_Release|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Windows_Debug|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Windows_Release|AnyCPU' " />
+  <ItemGroup>
+    <Compile Include="System\Security\Cryptography\Aes.cs" />
+    <Compile Include="System\Security\Cryptography\Helpers.cs" />
+    <Compile Include="Internal\Cryptography\AesImplementation.cs" />
+    <Compile Include="Internal\Cryptography\AesNativeCryptoTransform.cs" />
+    <Compile Include="Internal\Cryptography\AesCngCryptoDecryptor.cs" />
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Encryption.Aes/src/System.Security.Cryptography.Encryption.Aes.csproj]: These Cng files should all be Windows-only, right? (maybe the Native one, too?)

File: src/Common/src/Interop/Unix/libcrypto/Interop.EVP.Cipher.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+using System.Runtime.InteropServices;
+
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class libcrypto
+    {
+        // This was computed by sizeof(EVP_CIPHER_CTX) on a 64-bit Ubuntu
+        // machine running OpenSSL 1.0.1f.  If we end up making a native
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libcrypto/Interop.EVP.Cipher.cs]: "If we end up making"... we have that now, right?  Do we have an issue tracking moving whatever from here you want moved?

File: src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesCngCryptoDecryptor.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+            base.Dispose(disposing);
+        }
+
+        protected sealed override void Reset()
+        {
+            base.Reset();
+            if (_heldoverCipher != null)
+            {
+                Array.Clear(_heldoverCipher, 0, _heldoverCipher.Length);
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesCngCryptoDecryptor.cs]: i don't know the answer to this, but there's an ArrayT<T>.Clear in the ArrayT.cs file that a bunch of projects bring in...  We should follow-up to see what the rhyme or reason is to when to use that vs Array.Clear as is done here.

File: src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesCngCryptoTransform.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                Buffer.BlockCopy(_iv, 0, _currentIv, 0, _iv.Length);
+            }
+        }
+
+        protected abstract int UncheckedTransformBlock(SafeKeyHandle hKey, byte[] currentIv, byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset);
+
+        protected abstract byte[] UncheckedTransformFinalBlock(SafeKeyHandle hKey, byte[] currentIv, byte[] inputBuffer, int inputOffset, int inputCount);
+
+        private SafeKeyHandle _hKey;
+        private byte[] _iv;         // This IV never changes.
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesCngCryptoTransform.cs]: > This IV never changes

Kind of a strange comment when it's being change in Dispose above :smile:

File: src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesImplementation.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        public sealed override ICryptoTransform CreateEncryptor(byte[] rgbKey, byte[] rgbIV)
+        {
+            return CreateTransform(rgbKey, rgbIV, encrypting: true);
+        }
+
+        public sealed override void GenerateIV()
+        {
+            byte[] iv = new byte[this.BlockSize / 8];
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesImplementation.cs]: I assume the 8 here is bits-per-byte?  Might be nice to have it in a named constant.

File: src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesOpenSslCryptoTransform.cs [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+                Buffer.BlockCopy(inputBuffer, inputOffset, decrypt, _decryptBuffer.Length, inputCount);
+
+                _decryptBuffer = null;
+            }
+
+            byte[] rawDecrypt = ProcessFinalBlock(decrypt, decryptOffset, decryptCount);
+
+            if (rawDecrypt.Length == 0)
+            {
+                return new byte[0];
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesOpenSslCryptoTransform.cs]: Can we use ```Array.Empty<byte>()```?

File: src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesOpenSslCryptoTransform.cs [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+            outputBytes += bytesWritten;
+
+            if (outputBytes == output.Length)
+            {
+                return output;
+            }
+
+            if (outputBytes == 0)
+            {
+                return new byte[0];
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesOpenSslCryptoTransform.cs]: Ditto.

**Comment by stephentoub** [comment]
Other than the build issue and some minor comments, LGTM.

File: src/System.Security.Cryptography.Encryption.Aes/src/System.Security.Cryptography.Encryption.Aes.csproj (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Encryption.Aes/src/System.Security.Cryptography.Encryption.Aes.csproj]: Huh, I wonder why it seemed to be building legitimately on my machine.  Building Windows first must have somehow put it into a weird state where things magically worked.

Fixed.

File: src/Common/src/Interop/Unix/libcrypto/Interop.EVP.Cipher.cs (hunk 2 above)
Comment (bartonjs) [inline src/Common/src/Interop/Unix/libcrypto/Interop.EVP.Cipher.cs]: We do now, https://github.com/dotnet/corefx/issues/1963

File: src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesImplementation.cs (hunk 5 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesImplementation.cs]: Done.  While I was there I noticed that there are a few extraneous "this."s floating around, so I got rid of them, too.

File: src/System.Security.Cryptography.Encryption.Aes/src/System.Security.Cryptography.Encryption.Aes.csproj (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Encryption.Aes/src/System.Security.Cryptography.Encryption.Aes.csproj]: (AesNativeCryptoTransform is shared between Windows/CNG and Unix/OpenSSL)

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesImplementation.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        protected sealed override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+        }
+
+        private ICryptoTransform CreateTransform(byte[] rgbKey, byte[] rgbIV, bool encrypting)
+        {
+            if (rgbKey == null)
+                throw new ArgumentNullException("key");
+            int keySize = rgbKey.Length * BitsPerByte;
```
Comment (rielz) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesImplementation.cs]: This can overflow (See #1755)

Comment (bartonjs) [inline src/System.Security.Cryptography.Encryption.Aes/src/Internal/Cryptography/AesImplementation.cs]: Thanks for the feedback.  I've opened https://github.com/dotnet/corefx/issues/1981 for followup, but am going to merge as-is to best facilitate opening the library (given that this is an existing issue in the published package).

**Pull request merged**